### PR TITLE
[json-schema-validator] Conan v2 support

### PR DESCRIPTION
**Pull request by garethsb**
* Conan v2 support
* Bump nlohmann_json/3.11.2

### CODE REVIEW

**Comment by uilianries** [comment]
The CI linter in not happy. This recipe is outdated and requires more update. Could you please update it? We can help you. Please, take a look on the annotations on https://github.com/conan-io/conan-center-index/pull/13959/files. Also, https://github.com/conan-io/conan-center-index/blob/master/docs/v2_linter.md is a good guide.

File: recipes/json-schema-validator/all/conanfile.py [hunk 1]
```diff
@@ -94,20 +94,19 @@ def package(self):
...
 
         # TODO: to remove in conan v2 once cmake_find_package* generators removed
         self._create_cmake_module_alias_targets(
             os.path.join(self.package_folder, self._module_file_rel_path),
             {"nlohmann_json_schema_validator": "nlohmann_json_schema_validator::nlohmann_json_schema_validator"}
         )
 
-    @staticmethod
-    def _create_cmake_module_alias_targets(module_file, targets):
+    def _create_cmake_module_alias_targets(self, module_file, targets):
```
Comment (uilianries) [inline recipes/json-schema-validator/all/conanfile.py:109]:
```suggestion
    def _create_cmake_module_alias_targets(conanfile, module_file, targets):
```
I would prefer `conanfile` or anything else which is not `self`. As it's a static method, you are shadowing `self` using it as parameter name.

File: recipes/json-schema-validator/all/conanfile.py [hunk 2]
```diff
@@ -116,16 +115,16 @@ def _create_cmake_module_alias_targets(module_file, targets):
                     set_property(TARGET {alias} PROPERTY INTERFACE_LINK_LIBRARIES {aliased})
                 endif()
             """.format(alias=alias, aliased=aliased))
-        tools.save(module_file, content)
+        files.save(self, module_file, content)
```
Comment (uilianries) [inline recipes/json-schema-validator/all/conanfile.py:118]:
```suggestion
        files.save(conanfile, module_file, content)
```

File: recipes/json-schema-validator/all/conanfile.py (hunk 1 above)
Comment (garethsb) [reply recipes/json-schema-validator/all/conanfile.py:109]: Notice I removed the `@staticmethod` annotation...

Comment (uilianries) [reply recipes/json-schema-validator/all/conanfile.py:109]: Sorry, I misread the diff.

**Comment by garethsb** [comment]
@uilianries I added the patches to fix the `cmake_minimum_version` problem and now the linter is complaining about missing `patch_description` and `patch_type` entries... The docs for [`conan.tools.files.apply_conandata_patches()`](https://docs.conan.io/en/latest/reference/conanfile/tools/files/patches.html?highlight=patch_type#conan-tools-files-apply-conandata-patches) mention them in an example but otherwise the spec seems to be "**kwargs**: Extra parameters that can be added and will contribute to output information" - certainly nothing to imply they are required, and no list of valid `patch_type` values...

File: recipes/json-schema-validator/all/conanfile.py [hunk 3]
```diff
@@ -47,14 +42,14 @@ def configure(self):
...
     def requirements(self):
-        self.requires("nlohmann_json/3.10.5")
+        self.requires("nlohmann_json/3.11.2")
 
     def validate(self):
-        version = tools.Version(self.version)
+        version = scm.Version(self.version)
         min_vs_version = "16" if version < "2.1.0" else "14"
         min_cppstd = "17" if self.settings.compiler == "Visual Studio" and version < "2.1.0" else "11"
         if self.settings.get_safe("compiler.cppstd"):
```
Comment (uilianries) [inline recipes/json-schema-validator/all/conanfile.py:51]:
```suggestion
        if self.info.settings.get_safe("compiler.cppstd"):
```
https://docs.conan.io/en/latest/migrating_to_2.0/recipes.html#the-validate-method

File: recipes/json-schema-validator/all/conanfile.py [hunk 4]
```diff
@@ -64,50 +59,51 @@ def validate(self):
             "apple-clang": "9"}
         min_version = compilers.get(str(self.settings.compiler))
```
Comment (uilianries) [inline recipes/json-schema-validator/all/conanfile.py:60]:
```suggestion
        min_version = compilers.get(str(self.info.settings.compiler))
```

File: recipes/json-schema-validator/all/conanfile.py [hunk 5]
```diff
@@ -64,50 +59,51 @@ def validate(self):
...
         if not min_version:
-            self.output.warn("{} recipe lacks information about the {} compiler support.".format(
-                self.name, self.settings.compiler))
+            self.output.warn(f"{self.name} recipe lacks information about the {self.settings.compiler} compiler support.")
```
Comment (uilianries) [inline recipes/json-schema-validator/all/conanfile.py:62]:
```suggestion
            self.output.warn(f"{self.name} recipe lacks information about the {self.info.settings.compiler} compiler support.")
```

File: recipes/json-schema-validator/all/conanfile.py [hunk 6]
```diff
@@ -64,50 +59,51 @@ def validate(self):
...
         else:
-            if tools.Version(self.settings.compiler.version) < min_version:
-                raise ConanInvalidConfiguration("{} requires c++{} support. The current compiler {} {} does not support it.".format(
-                    self.name, min_cppstd, self.settings.compiler, self.settings.compiler.version))
+            if scm.Version(self.settings.compiler.version) < min_version:
```
Comment (uilianries) [inline recipes/json-schema-validator/all/conanfile.py:64]:
```suggestion
            if scm.Version(self.info.settings.compiler.version) < min_version:
```

File: recipes/json-schema-validator/all/conanfile.py [hunk 7]
```diff
@@ -64,50 +59,51 @@ def validate(self):
...
+                raise ConanInvalidConfiguration(f"{self.name} requires c++{min_cppstd} support. The current compiler {self.settings.compiler} {self.settings.compiler.version} does not support it.")
```
Comment (uilianries) [inline recipes/json-schema-validator/all/conanfile.py:65]:
```suggestion
                raise ConanInvalidConfiguration(f"{self.name} requires c++{min_cppstd} support. The current compiler {self.info.settings.compiler} {self.info.settings.compiler.version} does not support it.")
```

**Review by uilianries — COMMENTED** [review]
Looking good, need some adjusts

**Comment by uilianries** [comment]
> @uilianries I added the patches to fix the cmake_minimum_version problem and now the linter is complaining about missing patch_description and patch_type entries... The docs for [conan.tools.files.apply_conandata_patches()](https://docs.conan.io/en/latest/reference/conanfile/tools/files/patches.html?highlight=patch_type#conan-tools-files-apply-conandata-patches)

@garethsb It's not mandatory, but very welcome. Patches need to be described to help other users to understand their nature. More info: https://github.com/conan-io/conan-center-index/blob/master/docs/conandata_yml_format.md#patches-fields

**Comment by garethsb** [comment]
> More info: https://github.com/conan-io/conan-center-index/blob/master/docs/conandata_yml_format.md#patches-fields

@uilianries That's the docs I was looking for, thanks! I forget sometimes it's conan docs I want and sometimes CCI docs...

File: recipes/json-schema-validator/all/conanfile.py [hunk 8]
```diff
@@ -1,9 +1,11 @@
-from conans import ConanFile, CMake, tools
-from conans.errors import ConanInvalidConfiguration
+from conan import ConanFile
+from conan.errors import ConanInvalidConfiguration
+from conan.tools import build, files, scm
```
Comment (garethsb) [inline recipes/json-schema-validator/all/conanfile.py:3]:
```suggestion
from conan.tools import build, files, microsoft, scm
```

**Comment by uilianries** [comment]
> > More info: https://github.com/conan-io/conan-center-index/blob/master/docs/conandata_yml_format.md#patches-fields
> 
> @uilianries That's the docs I was looking for, thanks! I forget sometimes it's conan docs I want and sometimes CCI docs...

Everything is there, sometimes is hard to find or because the search engine is naive or because is more hidden than it should. In case you find some difficult to find anything there, please, open an issue to docs. For me and other devs, we used to read that everyday so we forget how hard can be finding some useful information there.

**Review by prince-chrismc — APPROVED** [review]

File: recipes/json-schema-validator/all/conanfile.py [hunk 9]
```diff
@@ -116,16 +111,16 @@ def _create_cmake_module_alias_targets(module_file, targets):
...
+        files.save(self, module_file, content)
 
     @property
     def _module_file_rel_path(self):
-        return os.path.join("lib", "cmake", "conan-official-{}-targets.cmake".format(self.name))
+        return os.path.join("lib", "cmake", f"conan-official-{self.name}-targets.cmake")
 
     def package_info(self):
         self.cpp_info.set_property("cmake_file_name", "nlohmann_json_schema_validator")
         self.cpp_info.set_property("cmake_target_name", "nlohmann_json_schema_validator")
```
Comment (garethsb) [inline recipes/json-schema-validator/all/conanfile.py:122]:
```suggestion
        self.cpp_info.set_property("cmake_target_name", "nlohmann_json_schema_validator::nlohmann_json_schema_validator")
```

**Review by uilianries — DISMISSED** [review]
LGTM

**Comment by garethsb** [comment]
@uilianries Do we need to upgrade the `test_package` to use `CMakeDeps` too? And do we have to keep the existing one as `test_v1_package`?

**Review by prince-chrismc — APPROVED** [review]

**Comment by uilianries** [comment]
> @uilianries Do we need to upgrade the `test_package` to use `CMakeDeps` too? And do we have to keep the existing one as `test_v1_package`?

Yes,, please.

**Review by uilianries — APPROVED** [review]
LGTM

File: recipes/json-schema-validator/all/conanfile.py [hunk 10]
```diff
@@ -116,16 +111,16 @@ def _create_cmake_module_alias_targets(module_file, targets):
...
     @property
     def _module_file_rel_path(self):
-        return os.path.join("lib", "cmake", "conan-official-{}-targets.cmake".format(self.name))
+        return os.path.join("lib", "cmake", f"conan-official-{self.name}-targets.cmake")
 
     def package_info(self):
         self.cpp_info.set_property("cmake_file_name", "nlohmann_json_schema_validator")
-        self.cpp_info.set_property("cmake_target_name", "nlohmann_json_schema_validator")
-        self.cpp_info.libs = ["json-schema-validator" if tools.Version(self.version) < "2.1.0" else "nlohmann_json_schema_validator"]
+        self.cpp_info.set_property("cmake_target_name", "nlohmann_json_schema_validator::nlohmann_json_schema_validator")
```
Comment (SpaceIm) [inline recipes/json-schema-validator/all/conanfile.py:122]: it's a regression, target name was correct and aligned with upstream imported target.

Comment (garethsb) [reply recipes/json-schema-validator/all/conanfile.py:122]: Thanks for noticing.

**Pull request merged**